hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieCDCLogger.java

```diff
@@ -0,0 +1,263 @@
+              oldRecord, record);
+        }
+      } else {
+        // deleted cdc record
+        cdcRecord = createCDCRecord(HoodieCDCOperation.DELETE, recordKey,
+            oldRecord, null);
+      }
+      cdcData.put(recordKey, new HoodieAvroPayload(Option.of(cdcRecord)));
+    }
+  }
+
+  private GenericData.Record createCDCRecord(HoodieCDCOperation operation,
+                                             String recordKey,
+                                             GenericRecord oldRecord,
+                                             GenericRecord newRecord) {
+    GenericData.Record record;
+    if (cdcSupplementalLoggingMode.equals(HoodieCDCSupplementalLoggingMode.WITH_BEFORE_AFTER)) {
+      record = HoodieCDCUtils.cdcRecord(operation.getValue(), commitTime,
+          removeCommitMetadata(oldRecord), newRecord);
+    } else if (cdcSupplementalLoggingMode.equals(HoodieCDCSupplementalLoggingMode.WITH_BEFORE)) {
+      record = HoodieCDCUtils.cdcRecord(operation.getValue(), recordKey,
+          removeCommitMetadata(oldRecord));
+    } else {
+      record = HoodieCDCUtils.cdcRecord(operation.getValue(), recordKey);
+    }
+    return record;
+  }
+
+  private GenericRecord removeCommitMetadata(GenericRecord record) {
+    if (record != null && populateMetaFields) {
+      GenericData.Record newRecord = new GenericData.Record(dataSchema);
+      for (Schema.Field field : dataSchema.getFields()) {
+        newRecord.put(field.name(), record.get(field.name()));
+      }
+      return newRecord;
+    }
+    return record;
+  }
+
+  public boolean isEmpty() {
+    return !this.cdcEnabled || this.cdcData.isEmpty();
+  }
+
+  public Option<AppendResult> writeCDCData() {
+    if (isEmpty()) {
+      return Option.empty();
+    }
+    try {
+      Map<HoodieLogBlock.HeaderMetadataType, String> header = buildCDCBlockHeader();
+      List<IndexedRecord> records = cdcData.values().stream()
+          .map(record -> {
+            try {
+              return record.getInsertValue(cdcSchema).get();
+            } catch (IOException e) {
+              throw new HoodieIOException("Failed to get cdc record", e);
+            }
+          }).collect(Collectors.toList());
+      HoodieLogBlock block = new HoodieCDCDataBlock(records, header, keyField);
+      AppendResult result = cdcWriter.appendBlocks(Collections.singletonList(block));
+
+      // call close to trigger the data flush.
+      this.close();
+
+      return Option.of(result);
+    } catch (Exception e) {
+      throw new HoodieException("Failed to write the cdc data to " + cdcWriter.getLogFile().getPath(), e);
+    }
+  }
+
+  private Map<HoodieLogBlock.HeaderMetadataType, String> buildCDCBlockHeader() {
+    Map<HoodieLogBlock.HeaderMetadataType, String> header = new HashMap<>();
+    header.put(HoodieLogBlock.HeaderMetadataType.INSTANT_TIME, commitTime);
+    if (cdcSupplementalLoggingMode.equals(HoodieCDCSupplementalLoggingMode.WITH_BEFORE_AFTER)) {
```

Comment:
nit: We can reduce conditional to pick the schema, then insert into header after

Reply:
done.